File: RadixCyclicNN/radixnet/encoding.py
"""How a text becomes the grams the graph is built from, and how it comes back.

:class:`Encoding` is three dials - what one *unit* of text is, how many units a
gram holds, and how far apart consecutive grams start:

* ``Encoding()`` is the default and the one the model was born with: character
  trigrams, stride 1 (``"hello" -> ["hel", "ell", "llo"]``).
* ``Encoding(n=5)`` is a sliding window of five characters, and any n works.
* ``Encoding(n=4, stride=4)`` is *tokenisation*: non-overlapping groups of four
  letters. So is ``Encoding(n=5, stride=5)``.
* ``Encoding(unit=WORDS, n=2)`` is the word bigram, ``n=3`` the word trigram.

:class:`Encoder` and :class:`Decoder` are the two halves of it kept as objects,
because that is how the rest of the package holds them; ``Encoder(window=5)``
still means what it always did.  :meth:`Decoder.decode_path` reverses the
(possibly path-compressed) node labels of the graph.

:func:`repair_base64` is the shared tail of the *media* text formats
(``img:...`` in :mod:`radixnet.vision`, ``aud:...`` in :mod:`radixnet.speech`):
the base64 payload of a text the network predicted is rarely clean, so it is
repaired before it is decoded.
"""

from __future__ import annotations

import base64
import binascii
import re
from collections.abc import Iterable, Sequence
from dataclasses import dataclass

WINDOW = 3
"""The default n of the n-gram: the trigram the model was born with."""

CHARS = "char"
"""One unit is one character."""
WORDS = "word"
"""One unit is one whitespace-delimited word; the text is normalised to single spaces."""
UNIT_KINDS = (CHARS, WORDS)

START_LABEL = "<s>"
END_LABEL = "</s>"
BACK_LABEL = "<back>"
"""The third sentinel: where the graph has learned that a walk goes round (``graph.BACK``)."""

# A "unit view" is whatever slices by unit: a str for characters (Python slices
# strings by code point already), a list of words for words.  ``len`` is the
# unit count of both, and :func:`_piece` writes a slice of either back out.
UnitView = "str | list[str]"


def _piece(view: str | Sequence[str], lo: int, hi: int | None = None) -> str:
    """Units ``[lo:hi)`` of a view, as text."""
    part = view[lo:hi]
    return part if isinstance(part, str) else " ".join(part)


@dataclass(frozen=True)
class Encoding:
    """How a text becomes grams, and how labels become text again.

    Three dials, fixed for a graph's life (every label, every index key and
    every offset is measured in the units of the encoding that built it):

    * ``unit`` - :data:`CHARS` or :data:`WORDS`.
    * ``n`` - units per gram: the *n* of the n-gram, any ``n >= 1``.
    * ``stride`` - units between two consecutive grams.  ``1`` slides the
      window, so consecutive grams share ``n - 1`` units, which is what lets
      the graph chain them; ``n`` cuts the text into groups that share nothing.
    """

    unit: str = CHARS
    n: int = WINDOW
    stride: int = 1

    def __post_init__(self) -> None:
        object.__setattr__(self, "unit", str(self.unit))
        object.__setattr__(self, "n", int(self.n))
        object.__setattr__(self, "stride", int(self.stride))
        self.validate()

    def validate(self) -> None:
        """Raise :class:`ValueError` if this is not an encoding the graph can be built in."""
        if self.unit not in UNIT_KINDS:
            raise ValueError(f"unit must be one of {UNIT_KINDS}, got {self.unit!r}")
        if self.n < 1:
            raise ValueError(f"n must be >= 1, got {self.n}")
        if self.stride < 1:
            raise ValueError(f"stride must be >= 1, got {self.stride}")
        if self.stride > self.n:
            raise ValueError(f"stride {self.stride} must be <= n {self.n}: a larger stride would skip units")

    # -- what it is ----------------------------------------------------------

    @property
    def overlap(self) -> int:
        """Units two consecutive grams share: ``n - stride``, so ``n - 1`` sliding and 0 in groups."""
        return self.n - self.stride

    @property
    def sliding(self) -> bool:
        """Do consecutive grams overlap at all?"""
        return self.stride < self.n

    def is_default(self) -> bool:
        """The character trigram of stride 1: what a model file leaves unwritten."""
        return self == Encoding()

    def __str__(self) -> str:
        return f"{self.unit}:{self.n}:{self.stride}"

    def describe(self) -> str:
        """The human form: ``'2-word grams, stride 1 (sliding)'``."""
        unit = "character" if self.unit == CHARS else "word"
        return f"{self.n}-{unit} grams, stride {self.stride} ({'sliding' if self.sliding else 'groups'})"

    def to_dict(self) -> dict:
        return {"unit": self.unit, "n": self.n, "stride": self.stride}

    @classmethod
    def from_dict(cls, d: dict | None) -> "Encoding":
        """Read an ``encoding`` block of a model file; ``None`` is the default."""
        if not d:
            return cls()
        return cls(unit=str(d.get("unit", CHARS)), n=int(d.get("n", WINDOW)), stride=int(d.get("stride", 1)))

    # -- units ---------------------------------------------------------------

    def units(self, text: str) -> str | list[str]:
        """The text as something that slices by unit: itself, or its words."""
        return text if self.unit == CHARS else text.split()

    def length(self, text: str) -> int:
        """How many units a text holds."""
        return len(text) if self.unit == CHARS else len(text.split())

    def piece(self, text: str, lo: int, hi: int | None = None) -> str:
        """Units ``[lo:hi)`` of a text, as text."""
        return _piece(self.units(text), lo, hi)

    def join(self, *parts: str) -> str:
        """Glue unit-aligned pieces: nothing between characters, one space between words.

        Empty pieces are dropped, so a label contributing no unit adds no separator.
        """
        if self.unit != WORDS:
            return "".join(parts)
        return " ".join(p for p in parts if p)

    def truncate(self, text: str, n: int) -> str:
        """The first ``n`` units; ``n < 0`` leaves the text alone."""
        return text if n < 0 else self.piece(text, 0, n)

    def has_unit_prefix(self, text: str, prefix: str) -> bool:
        """Does ``text`` start with ``prefix`` *on a unit boundary*?

        For words that means whole words: ``"the ca"`` is not a prefix of
        ``"the cat sat"``, ``"the cat"`` is.
        """
        if not prefix:
            return True
        if self.unit != WORDS:
            return text.startswith(prefix)
        return text == prefix or text.startswith(prefix + " ")

    # -- encoder -------------------------------------------------------------

    def covered(self, units: int) -> int:
        """How many of ``units`` units the grams of a text actually cover."""
        if units < self.n:
            return 0
        return (units - self.n) // self.stride * self.stride + self.n

    def encode(self, text: str) -> list[str]:
        """The grams of a text: ``n`` units each, ``stride`` units apart; ``[]`` if it is too short.

        The tail that does not fill a whole gram is dropped, exactly as the
        trigram encoding drops the last two characters of a text.
        :meth:`normalize` is what is left.
        """
        view = self.units(text)
        last = len(view) - self.n
        if last < 0:
            return []
        return [_piece(view, i, i + self.n) for i in range(0, last + 1, self.stride)]

    def normalize(self, text: str) -> str:
        """The text this encoding can represent - what a round trip through the graph returns.

        For the default encoding that is the text itself; a word encoding loses
        the original spacing, and a grouping one the tail that does not fill a
        group.
        """
        view = self.units(text)
        covered = self.covered(len(view))
        return _piece(view, 0, covered) if covered else ""

    # -- decoder -------------------------------------------------------------

    def decode_grams(self, grams: Iterable[str]) -> str:
        """The inverse of :meth:`encode`, up to the tail it dropped."""
        grams = list(grams)
        if not grams:
            return ""
        return self.join(grams[0], *(self.piece(g, self.overlap) for g in grams[1:]))

    def decode_path(
        self,
        labels: Iterable[str],
        start_offset: int = 0,
        include_context: bool = True,
        skip_sentinels: bool = True,
    ) -> str:
        """Decode node labels in path order into text (see :meth:`Decoder.decode_path`)."""
        if start_offset < 0:
            raise ValueError(f"start_offset must be >= 0, got {start_offset}")
        first_cut = start_offset if include_context else start_offset + self.n
        parts: list[str] = []
        first = True
        for label in labels:
            if skip_sentinels and label in (START_LABEL, END_LABEL, BACK_LABEL):
                continue
            parts.append(self.piece(label, first_cut if first else self.overlap))
            first = False
        return self.join(*parts)


DEFAULT_ENCODING = Encoding()
"""Character trigrams of stride 1: what every model used before the encoding became a choice."""

_ALIASES = {
    "": Encoding(),
    "default": Encoding(),
    "trigram": Encoding(),
    "trigrams": Encoding(),
    "bigram": Encoding(n=2),
    "bigrams": Encoding(n=2),
    "word": Encoding(unit=WORDS, n=1),
    "words": Encoding(unit=WORDS, n=1),
    "word-unigram": Encoding(unit=WORDS, n=1),
    "word-bigram": Encoding(unit=WORDS, n=2),
    "word-bigrams": Encoding(unit=WORDS, n=2),
    "word-trigram": Encoding(unit=WORDS, n=3),
    "word-trigrams": Encoding(unit=WORDS, n=3),
}
_UNIT_NAMES = {
    "char": CHARS, "chars": CHARS, "character": CHARS, "characters": CHARS, "letter": CHARS, "letters": CHARS,
    "word": WORDS, "words": WORDS,
}


def parse_encoding(spec: str) -> Encoding:
    """Read a spec: ``unit[:n[:stride]]``, or one of the names in :data:`_ALIASES`.

    ``"char:3:1"`` is the default, ``"char:5:groups"`` (or ``"char:5:5"``)
    non-overlapping groups of five letters, ``"word:2"`` the word bigram.
    """
    text = (spec or "").strip().lower()
    if text in _ALIASES:
        return _ALIASES[text]
    parts = text.split(":")
    if len(parts) > 3:
        raise ValueError(f"encoding {spec!r}: expected unit[:n[:stride]]")
    if parts[0] not in _UNIT_NAMES:
        raise ValueError(f"encoding {spec!r}: unit must be char or word, got {parts[0]!r}")
    unit = _UNIT_NAMES[parts[0]]
    n = WINDOW
    if len(parts) > 1 and parts[1]:
        try:
            n = int(parts[1])
        except ValueError:
            raise ValueError(f"encoding {spec!r}: n must be a number, got {parts[1]!r}") from None
    stride = 1
    if len(parts) > 2 and parts[2]:
        if parts[2] in ("groups", "group", "blocks", "block"):
            stride = n
        elif parts[2] in ("sliding", "slide"):
            stride = 1
        else:
            try:
                stride = int(parts[2])
            except ValueError:
                raise ValueError(f"encoding {spec!r}: stride must be a number, got {parts[2]!r}") from None
    try:
        return Encoding(unit=unit, n=n, stride=stride)
    except ValueError as exc:
        raise ValueError(f"encoding {spec!r}: {exc}") from None


class Encoder:
    """The encoder half of an :class:`Encoding`, kept as an object.

    ``Encoder()`` is the sliding window of three characters; ``Encoder(5)`` a
    window of five; ``Encoder(encoding=Encoding(unit=WORDS, n=2))`` the word
    bigram.
    """

    __slots__ = ("encoding",)

    def __init__(self, window: int = WINDOW, encoding: Encoding | None = None) -> None:
        self.encoding = encoding if encoding is not None else Encoding(n=window)

    @property
    def window(self) -> int:
        """The n of the n-gram (the name this had when three was the only answer)."""
        return self.encoding.n

    def encode(self, text: str) -> list[str]:
        """Return the grams of ``text``; ``[]`` if it holds fewer than ``n`` units."""
        return self.encoding.encode(text)

    def __repr__(self) -> str:
        return f"Encoder(encoding={self.encoding})"


class Decoder:
    """Inverse of :class:`Encoder`, for raw grams and for compressed node labels."""

    __slots__ = ("encoding",)

    def __init__(self, window: int = WINDOW, encoding: Encoding | None = None) -> None:
        self.encoding = encoding if encoding is not None else Encoding(n=window)

    @property
    def window(self) -> int:
        """The n of the n-gram."""
        return self.encoding.n

    def decode_trigrams(self, grams: Iterable[str]) -> str:
        """``[] -> ""``; the first gram in full, then the part of each following one past the overlap."""
        return self.encoding.decode_grams(grams)

    def decode_path(
        self,
        labels: Iterable[str],
        start_offset: int = 0,
        include_context: bool = True,
        skip_sentinels: bool = True,
    ) -> str:
        """Decode node labels in path order into text.

        ``START_LABEL`` / ``END_LABEL`` entries are skipped while
        ``skip_sentinels`` is true.  Every label after the first *real* one
        contributes the part of it past the overlap (``n - stride`` units).
        The first real label contributes

        * ``label[start_offset:]`` when ``include_context`` is true (full
          context, used when generating from START), or
        * ``label[start_offset + n:]`` when it is false: the deterministic
          remainder of a compressed node *after* the matched gram, so a
          prediction returns only the continuation.

        A real node can legitimately carry a label equal to a sentinel label
        (the text ``"x<s>y"`` contains the trigram ``"<s>"``).  Callers that
        know the node ids strip START/END themselves and pass
        ``skip_sentinels=False`` so such a node is decoded like any other.
        """
        return self.encoding.decode_path(labels, start_offset, include_context, skip_sentinels)

    def __repr__(self) -> str:
        return f"Decoder(encoding={self.encoding})"


_B64_JUNK = re.compile(r"[^A-Za-z0-9+/=]")


def repair_base64(body: str) -> tuple[bytes, bool]:
    """Decode the base64 tail of a media text, repairing it first; ``(payload, repaired)``.

    The media encoders (:mod:`radixnet.vision`, :mod:`radixnet.speech`) pack
    their payload as base64 into a text the network trains on and *predicts*,
    so what comes back may be cut off, padded with junk or interrupted by
    whitespace.  Characters outside the base64 alphabet are dropped, a single
    dangling character (which can never decode) is removed with them, the
    padding is completed, and ``repaired`` says whether any of that changed
    the text.  The payload is returned as it decodes - callers pad or truncate
    it to the length their format needs.
    """
    clean = _B64_JUNK.sub("", body).rstrip("=")
    if len(clean) % 4 == 1:  # a single dangling character can never decode
        clean = clean[:-1]
    padded = clean + "=" * (-len(clean) % 4)
    repaired = padded != body.strip()  # a clean text comes back unchanged, padding included
    try:
        payload = base64.b64decode(padded, validate=True)
    except (ValueError, binascii.Error) as exc:  # pragma: no cover - the junk filter makes this rare
        raise ValueError(f"the base64 part cannot be decoded: {exc}") from exc
    return payload, repaired

File: RadixCyclicNN/go/radixnet/encoding.go
// Package radixnet is the Go port of the count / reward model of RadixCyclicNN:
// a self-compressing cyclic n-gram graph whose edge weights are a dual
// frequency function of traversal counts (all time and inside a sliding
// window) plus rewards, with beam-search prediction (top-K and bottom-K
// continuations), generation, scoring, 2NRL feedback and a self-conversation.
//
// Model files are interchangeable with the Python implementation
// (format "radixnet-count"): both sides read and write the same JSON layout,
// including the Mersenne Twister state, so a model trained here continues in
// Python and vice versa with identical numbers - in every encoding, not only
// the default one.  A graph built with anything but the character trigram of
// stride 1 writes an "encoding" block, and both implementations read it.
//
// Concurrency: training fans goroutines out over the texts (lines, paragraphs
// or pages of a file) for encoding, tracing and counting (lock-free atomic
// increments), recomputes weights and edge costs in parallel over the nodes,
// and runs the two beams of a prediction side by side.  Only structural
// changes (splits and merges of nodes) take a write lock.
package radixnet

import (
	"encoding/base64"
	"fmt"
	"regexp"
	"strconv"
	"strings"
	"unicode/utf8"
)

// Window is the default n-gram size: the trigram the model was born with.  It
// is the default of Encoding.N, not a property of every graph - a graph
// encodes with its own Encoding (Graph.Enc), whose n can be any number.
const Window = 3

// Overlap is how many units two consecutive labels share under the default
// encoding (a sliding window of stride 1).  A graph's own overlap is
// Graph.Enc.Overlap().
const Overlap = Window - 1

// StartLabel and EndLabel are the labels of the two sentinel nodes.
const (
	StartLabel = "<s>"
	EndLabel   = "</s>"
	// BackLabel is the third sentinel: where the graph has learned a walk goes round (Back).
	BackLabel = "<back>"
)

// -- what a unit is --------------------------------------------------------------

// UnitKind is what one position of a text is: a character or a word.  It is
// the atom everything downstream counts in - the n of the n-gram, the stride,
// a node's label length, the length of a prediction, the spans of a diff.
type UnitKind string

const (
	// Chars: one unit is one character (one code point).  "hello" is 5 units.
	Chars UnitKind = "char"
	// Words: one unit is one whitespace-delimited word.  "the cat sat" is 3
	// units, and the text is normalised to single spaces on the way in - a
	// word encoding keeps the words, not the layout.
	Words UnitKind = "word"
)

// Valid reports whether k is a unit kind this package knows.
func (k UnitKind) Valid() bool { return k == Chars || k == Words }

// -- the encoding ------------------------------------------------------------------

// Encoding is how a text becomes the grams the graph is built from, and how
// the labels of a walk become text again.  Three dials:
//
//   - Unit  - what one position is: a character (Chars) or a word (Words).
//   - N     - how many units one gram holds: the n of the n-gram.  Any n >= 1.
//   - Stride - how many units apart consecutive grams start.  1 is the sliding
//     window the model was born with ("hello" -> hel, ell, llo); N is
//     non-overlapping groups ("hello" with N=4 -> hell), and anything in
//     between overlaps by N - Stride.
//
// So the trigram model is {Chars, 3, 1}, groups of five letters are
// {Chars, 5, 5}, and word bigrams are {Words, 2, 1}.
//
// Two consecutive grams of a text share Overlap() units, which is what lets
// the graph chain them: node A ends with the units node B starts with.  With
// Stride == N nothing is shared, and the graph becomes a plain chain of
// groups - still a graph, just one whose nodes only ever meet at their ends.
type Encoding struct {
	Unit   UnitKind `json:"unit"`
	N      int      `json:"n"`
	Stride int      `json:"stride"`
}

// DefaultEncoding is the character trigram with stride 1: what every model
// used before the encoding became a choice, and the only encoding the Python
// implementation reads.
func DefaultEncoding() Encoding { return Encoding{Unit: Chars, N: Window, Stride: 1} }

// WithDefaults fills the zero value in: an empty unit is Chars, n 0 is Window,
// stride 0 is 1 (a sliding window).  Every entry point runs its encoding
// through this, so the zero Encoding is the default one.
func (e Encoding) WithDefaults() Encoding {
	if e.Unit == "" {
		e.Unit = Chars
	}
	if e.N == 0 {
		e.N = Window
	}
	if e.Stride == 0 {
		e.Stride = 1
	}
	return e
}

// Validate reports what is wrong with an encoding, if anything.
func (e Encoding) Validate() error {
	if !e.Unit.Valid() {
		return fmt.Errorf("unit must be %q or %q, got %q", Chars, Words, e.Unit)
	}
	if e.N < 1 {
		return fmt.Errorf("n must be >= 1, got %d", e.N)
	}
	if e.Stride < 1 {
		return fmt.Errorf("stride must be >= 1, got %d", e.Stride)
	}
	if e.Stride > e.N {
		return fmt.Errorf("stride %d must be <= n %d: a larger stride would skip units", e.Stride, e.N)
	}
	return nil
}

// Overlap is how many units two consecutive grams share: N - Stride, so
// N - 1 for a sliding window and 0 for groups.
func (e Encoding) Overlap() int { return e.N - e.Stride }

// IsDefault reports whether this is the character trigram of stride 1 - the
// encoding a model file leaves unwritten, and the one Python also speaks.
func (e Encoding) IsDefault() bool { return e == DefaultEncoding() }

// Sliding reports whether consecutive grams overlap at all.
func (e Encoding) Sliding() bool { return e.Stride < e.N }

// String is the compact spec ParseEncoding reads back: "char:3:1".
func (e Encoding) String() string { return fmt.Sprintf("%s:%d:%d", e.Unit, e.N, e.Stride) }

// Describe is the human form: "character trigrams, stride 1 (sliding)".
func (e Encoding) Describe() string {
	unit := "character"
	if e.Unit == Words {
		unit = "word"
	}
	kind := "sliding"
	if !e.Sliding() {
		kind = "groups"
	}
	return fmt.Sprintf("%d-%s grams, stride %d (%s)", e.N, unit, e.Stride, kind)
}

// ParseEncoding reads a spec: "unit[:n[:stride]]", where unit is char / chars /
// character / word / words, n defaults to Window and stride to 1 (a sliding
// window); "unit:n:n" and the shorthand "unit:n:groups" are the
// non-overlapping form.  A few names for the common ones are accepted too:
// "trigram", "bigram", "word-bigram", "word-trigram".
func ParseEncoding(spec string) (Encoding, error) {
	s := strings.ToLower(strings.TrimSpace(spec))
	if s == "" {
		return DefaultEncoding(), nil
	}
	switch s {
	case "default", "trigram", "trigrams":
		return DefaultEncoding(), nil
	case "bigram", "bigrams":
		return Encoding{Chars, 2, 1}, nil
	case "word", "words", "unigram-words", "word-unigram":
		return Encoding{Words, 1, 1}, nil
	case "word-bigram", "word-bigrams", "bigram-words":
		return Encoding{Words, 2, 1}, nil
	case "word-trigram", "word-trigrams", "trigram-words":
		return Encoding{Words, 3, 1}, nil
	}
	parts := strings.Split(s, ":")
	if len(parts) > 3 {
		return Encoding{}, fmt.Errorf("encoding %q: expected unit[:n[:stride]]", spec)
	}
	e := DefaultEncoding()
	switch parts[0] {
	case "char", "chars", "character", "characters", "letter", "letters":
		e.Unit = Chars
	case "word", "words":
		e.Unit = Words
	default:
		return Encoding{}, fmt.Errorf("encoding %q: unit must be char or word, got %q", spec, parts[0])
	}
	if len(parts) > 1 && parts[1] != "" {
		n, err := strconv.Atoi(parts[1])
		if err != nil {
			return Encoding{}, fmt.Errorf("encoding %q: n must be a number, got %q", spec, parts[1])
		}
		e.N = n
	}
	e.Stride = 1
	if len(parts) > 2 && parts[2] != "" {
		switch parts[2] {
		case "groups", "group", "blocks", "block":
			e.Stride = e.N
		case "sliding", "slide":
			e.Stride = 1
		default:
			st, err := strconv.Atoi(parts[2])
			if err != nil {
				return Encoding{}, fmt.Errorf("encoding %q: stride must be a number, got %q", spec, parts[2])
			}
			e.Stride = st
		}
	}
	if err := e.Validate(); err != nil {
		return Encoding{}, fmt.Errorf("encoding %q: %w", spec, err)
	}
	return e, nil
}

// -- units -------------------------------------------------------------------------

// Units is a text cut into units, indexed so that slicing by unit is O(1) -
// which is what the graph does all day (every split, every merge, every
// re-index of a label walks its grams).  starts holds the byte offset of every
// unit plus the end of the text.
type Units struct {
	text   string
	starts []int
	sep    int
}

// Len is the number of units.
func (u Units) Len() int { return len(u.starts) - 1 }

// Text is the whole (normalised) text.
func (u Units) Text() string { return u.text }

// Slice is units [from, to) as text; to < 0 means "to the end".  Out-of-range
// bounds are clamped, and an empty range is "".
func (u Units) Slice(from, to int) string {
	n := u.Len()
	if to < 0 || to > n {
		to = n
	}
	if from < 0 {
		from = 0
	}
	if from >= to {
		return ""
	}
	end := u.starts[to]
	if to < n {
		end -= u.sep // the separator before unit `to` belongs to neither side
	}
	return u.text[u.starts[from]:end]
}

// At is one unit.
func (u Units) At(i int) string { return u.Slice(i, i+1) }

// Units cuts text into the units of this encoding, normalising it on the way:
// a word encoding collapses every run of whitespace to a single space, a
// character encoding leaves the text exactly as it is.
func (e Encoding) Units(text string) Units {
	if e.Unit == Words {
		return wordUnits(text)
	}
	return charUnits(text)
}

func charUnits(text string) Units {
	starts := make([]int, 0, len(text)+1)
	for i := range text {
		starts = append(starts, i)
	}
	starts = append(starts, len(text))
	return Units{text: text, starts: starts, sep: 0}
}

func wordUnits(text string) Units {
	fields := strings.Fields(text)
	if len(fields) == 0 {
		return Units{text: "", starts: []int{0}, sep: 1}
	}
	var b strings.Builder
	size := len(fields) - 1
	for _, f := range fields {
		size += len(f)
	}
	b.Grow(size)
	starts := make([]int, 0, len(fields)+1)
	for i, f := range fields {
		if i > 0 {
			b.WriteByte(' ')
		}
		starts = append(starts, b.Len())
		b.WriteString(f)
	}
	starts = append(starts, b.Len())
	return Units{text: b.String(), starts: starts, sep: 1}
}

// Len is the number of units in a text.
func (e Encoding) Len(text string) int {
	if e.Unit == Words {
		return len(strings.Fields(text))
	}
	return utf8.RuneCountInString(text)
}

// Slice is units [from, to) of a text; to < 0 means "to the end".
func (e Encoding) Slice(text string, from, to int) string {
	if e.Unit == Chars {
		return runeSlice(text, from, to)
	}
	return e.Units(text).Slice(from, to)
}

// Join glues unit-aligned pieces back together: nothing between characters, a
// single space between words.  Empty pieces are dropped, so a label that
// contributes no unit adds no separator.
func (e Encoding) Join(parts ...string) string {
	if e.Unit != Words {
		return strings.Join(parts, "")
	}
	kept := parts[:0:0]
	for _, p := range parts {
		if p != "" {
			kept = append(kept, p)
		}
	}
	return strings.Join(kept, " ")
}

// Truncate is the first n units of a text; n < 0 leaves it alone.
func (e Encoding) Truncate(text string, n int) string {
	if n < 0 {
		return text
	}
	if e.Unit == Chars {
		return truncateRunes(text, n)
	}
	return e.Slice(text, 0, n)
}

// HasUnitPrefix reports whether text starts with prefix on a unit boundary:
// plain string prefix for characters, whole words for words ("the ca" is not a
// word prefix of "the cat sat", "the cat" is).
func (e Encoding) HasUnitPrefix(text, prefix string) bool {
	if prefix == "" {
		return true
	}
	if e.Unit != Words {
		return strings.HasPrefix(text, prefix)
	}
	return text == prefix || strings.HasPrefix(text, prefix+" ")
}

// -- encoder -----------------------------------------------------------------------

// Encode cuts text into its grams: N units each, Stride units apart.  nil when
// the text holds fewer than N units.
//
//	{Chars, 3, 1} "hello"       -> ["hel", "ell", "llo"]
//	{Chars, 4, 4} "hello there" -> ["hell", "o th", "ere"] -> ["hell", "o th"]
//	{Words, 2, 1} "the cat sat" -> ["the cat", "cat sat"]
//
// The tail that does not fill a whole gram is dropped, exactly as the last two
// characters of a text are dropped by the trigram encoding: Normalize says
// what is left.
func (e Encoding) Encode(text string) []string {
	e = e.WithDefaults()
	if e.Unit == Chars { // the hot path of a training pass: no unit index needed
		runes := []rune(text)
		n := len(runes)
		if n < e.N {
			return nil
		}
		grams := make([]string, (n-e.N)/e.Stride+1)
		for i := range grams {
			at := i * e.Stride
			grams[i] = string(runes[at : at+e.N])
		}
		return grams
	}
	return e.EncodeUnits(e.Units(text))
}

// EncodeUnits is Encode over an already-cut text, for callers that need both.
func (e Encoding) EncodeUnits(u Units) []string {
	n := u.Len()
	if n < e.N {
		return nil
	}
	count := (n-e.N)/e.Stride + 1
	grams := make([]string, count)
	for i := 0; i < count; i++ {
		at := i * e.Stride
		grams[i] = u.Slice(at, at+e.N)
	}
	return grams
}

// Covered is how many units of a text its grams actually cover: everything for
// a sliding window, everything but the ragged tail for groups.  0 when the
// text is too short to encode.
func (e Encoding) Covered(n int) int {
	if n < e.N {
		return 0
	}
	return (n-e.N)/e.Stride*e.Stride + e.N
}

// Normalize is the text this encoding can represent: the units its grams cover,
// written the way the decoder writes them.  For the default encoding that is
// the text itself; a word encoding loses the original spacing, and a grouping
// encoding loses the tail that does not fill a group.  It is what a round trip
// through the graph returns, so it is what a round trip is checked against.
func (e Encoding) Normalize(text string) string {
	e = e.WithDefaults()
	u := e.Units(text)
	covered := e.Covered(u.Len())
	if covered == 0 {
		return ""
	}
	return u.Slice(0, covered)
}

// -- decoder -----------------------------------------------------------------------

// DecodePath decodes node labels in path order into text.  Every label after
// the first contributes its part beyond the overlap; the first contributes
// label[startOffset:] when includeContext is true, else label[startOffset+N:]
// (the deterministic remainder of a compressed node after the matched gram).
// Offsets are in units, and sentinel labels are never skipped here: callers
// strip START / END by id.
func (e Encoding) DecodePath(labels []string, startOffset int, includeContext bool) string {
	e = e.WithDefaults()
	firstCut := startOffset
	if !includeContext {
		firstCut += e.N
	}
	if e.Unit == Chars {
		out := make([]byte, 0, 64)
		for i, label := range labels {
			cut := e.Overlap()
			if i == 0 {
				cut = firstCut
			}
			out = append(out, runeSlice(label, cut, -1)...)
		}
		return string(out)
	}
	parts := make([]string, 0, len(labels))
	for i, label := range labels {
		cut := e.Overlap()
		if i == 0 {
			cut = firstCut
		}
		parts = append(parts, e.Slice(label, cut, -1))
	}
	return e.Join(parts...)
}

// DecodeGrams decodes raw grams in order - the inverse of Encode, up to the
// tail Encode dropped.
func (e Encoding) DecodeGrams(grams []string) string {
	e = e.WithDefaults()
	parts := make([]string, 0, len(grams))
	for i, gram := range grams {
		if i == 0 {
			parts = append(parts, gram)
			continue
		}
		parts = append(parts, e.Slice(gram, e.Overlap(), -1))
	}
	return e.Join(parts...)
}

// -- the default encoding, for callers that have no graph to ask -------------------

// Encode is DefaultEncoding().Encode: the overlapping character trigrams of a
// text.  Code that belongs to a graph encodes with that graph's Encoding
// (Graph.Encode) instead.
func Encode(text string) []string { return DefaultEncoding().Encode(text) }

// DecodePath is DefaultEncoding().DecodePath.
func DecodePath(labels []string, startOffset int, includeContext bool) string {
	return DefaultEncoding().DecodePath(labels, startOffset, includeContext)
}

// -- rune helpers ------------------------------------------------------------------

// runeLen is the character (code point) length of a string, the unit the
// Python implementation measures text in.
func runeLen(s string) int { return utf8.RuneCountInString(s) }

// runeSlice returns s[from:to] in characters (code points); to < 0 means "to the end".
func runeSlice(s string, from, to int) string {
	if from <= 0 && to < 0 {
		return s
	}
	if from < 0 {
		from = 0
	}
	i := 0
	start, end := -1, len(s)
	for bi := range s {
		if i == from {
			start = bi
		}
		if to >= 0 && i == to {
			end = bi
			break
		}
		i++
	}
	if start < 0 {
		if from == i { // from == len(s) in characters
			return ""
		}
		return ""
	}
	if end < start {
		return ""
	}
	return s[start:end]
}

// DecodeTrigrams is the inverse of Encode under the default encoding: the
// first gram in full, then the part of each following one past the overlap.
// A graph decodes with its own encoding (Encoding.DecodeGrams).
func DecodeTrigrams(grams []string) string { return DefaultEncoding().DecodeGrams(grams) }

// truncateRunes returns the first n characters of s.
func truncateRunes(s string, n int) string {
	if n < 0 {
		return s
	}
	i := 0
	for bi := range s {
		if i == n {
			return s[:bi]
		}
		i++
	}
	return s
}

// b64Junk is everything outside the base64 alphabet: a predicted payload is
// routinely interrupted by whitespace or a stray character.
var b64Junk = regexp.MustCompile(`[^A-Za-z0-9+/=]`)

// RepairBase64 decodes the base64 tail of a media text, repairing it first;
// returns (payload, repaired, error).
//
// The media encoders (vision.go, speech.go) pack their payload as base64 into a
// text the network trains on and *predicts*, so what comes back may be cut off,
// padded with junk or interrupted by whitespace.  Characters outside the
// alphabet are dropped, a single dangling character (which can never decode)
// goes with them, the padding is completed, and `repaired` says whether any of
// that changed the text.  The payload comes back as it decodes - callers pad or
// truncate it to the length their format needs.
func RepairBase64(body string) ([]byte, bool, error) {
	clean := strings.TrimRight(b64Junk.ReplaceAllString(body, ""), "=")
	if len(clean)%4 == 1 { // a single dangling character can never decode
		clean = clean[:len(clean)-1]
	}
	padded := clean + strings.Repeat("=", (4-len(clean)%4)%4)
	repaired := padded != strings.TrimSpace(body) // a clean text comes back unchanged, padding included
	payload, err := base64.StdEncoding.DecodeString(padded)
	if err != nil {
		return nil, repaired, fmt.Errorf("the base64 part cannot be decoded: %w", err)
	}
	return payload, repaired, nil
}
